cloud-storage/cloud_storage.c

```diff
@@ -0,0 +1,947 @@
+/* Copyright  (C) 2020 The RetroArch team
+ *
+ * ---------------------------------------------------------------------------------------
+ * The following license statement only applies to this file (cloud_storage.c).
+ * ---------------------------------------------------------------------------------------
+ *
+ * Permission is hereby granted, free of charge,
+ * to any person obtaining a copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
+ * and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <stdarg.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <file/file_path.h>
+#include <retroarch.h>
+#include <retro_dirent.h>
+#include <utils/md5.h>
+#include <lists/linked_list.h>
+#include <queues/generic_queue.h>
+#include <rthreads/rthreads.h>
+#include <string/stdstring.h>
+#include <streams/file_stream.h>
+
+#include "cloud_storage.h"
+#include "provider_common.h"
+#include "providers/local_folder/local_folder.h"
+
+#define BUFFER_SIZE 8192
+
+enum _operation_type_t
+{
+   SYNC_FILES,
+   UPLOAD_FILE,
+   DOWNLOAD_FILE
+};
+
+struct _operation_t
+{
+   enum _operation_type_t operation_type;
+
+   union
+   {
+      struct
+      {
+         folder_type_t folder_type;
+      } sync;
+
+      struct
+      {
+         folder_type_t folder_type;
+         char *local_file;
+      } upload;
+
+      struct
+      {
+         folder_type_t folder_type;
+         cloud_storage_item_t *remote_file;
+      } download;
+   } parameters;
+};
+
+static bool _shutdown = true;
+static cloud_storage_provider_t *_provider;
+
+static char _game_states_config_value[8192] = "";
+
+static bool _need_sync_runtime_logs = false;
+
+cloud_storage_item_t *_game_states = NULL;
+
+static generic_queue_t *_operation_queue = NULL;
+static slock_t *_operation_queue_mutex;
+
+static scond_t *_operation_condition;
+static sthread_t *_operation_thread;
+
+static void _sync_files_execute(folder_type_t folder_type);
+
+static void _upload_file_execute(folder_type_t folder_type, char *file_name);
+
+void _cloud_storage_item_free_fn(void *item)
```

Comment:
If the goal is to make this function private, please add "static" so that it remains private within the compilation unit.

Reply:
Changed to be static. Good catch.